### PR TITLE
[dpdk] Emit a pseudo header, for dpdk instructions which require

### PR DESCRIPTION
**Pull request by kamleshbhalui**
Emit a pseudo header, for dpdk instructions which require it's operands to be in a header.
This patch does following:

create a pseudo header
move concerned instruction's operand to the header fields created in step 1
make the newly created header fields part of header.

### CODE REVIEW

**Comment by usha1830** [comment]
Fixes https://github.com/p4lang/p4c/issues/3676

File: backends/dpdk/dpdkArch.cpp [hunk 1]
```diff
@@ -2877,4 +2879,174 @@ const IR::Node* ElimHeaderCopy::postorder(IR::Member* m) {
...
+    IR::Type_Name* newTname = nullptr;
+    IR::StructExpression* struct_exp = nullptr;
+    auto result = new IR::IndexedVector<IR::StatOrDecl>();
+    if (auto* m = mce->method->to<IR::Member>()) {
+        if (auto* type = typeMap->getType(m->expr)->to<IR::Type_Extern>()) {
+            if (type->name == "InternetChecksum") {
+                if (m->member == "add" || m->member == "subtract") {
+                    auto components = new IR::IndexedVector<IR::NamedExpression>();
+                    for (auto arg : *mce->arguments) {
+                        BUG_CHECK(arg->expression->is<IR::StructExpression>(),
```
Comment (usha1830) [inline backends/dpdk/dpdkArch.cpp:2980]: I think there is no restriction on the arguments to be StructExpression. Constants, PathExpressions, Member expressions should be allowed as well.

File: backends/dpdk/dpdkArch.h [hunk 2]
```diff
@@ -1135,6 +1135,176 @@ class ElimHeaderCopy : public Transform {
...
+    HaveNonHeaderChecksumArgs(P4::TypeMap* typeMap, bool& is_all_arg_header_fields)
+        : typeMap(typeMap), is_all_arg_header_fields(is_all_arg_header_fields) {}
+    bool preorder(const IR::MethodCallExpression* mce) override {
+        if (!is_all_arg_header_fields) return false;
+        if (auto* m = mce->method->to<IR::Member>()) {
+            if (auto* type = typeMap->getType(m->expr)->to<IR::Type_Extern>()) {
+                if (type->name == "InternetChecksum") {
+                    if (m->member == "add" || m->member == "subtract") {
+                        for (auto arg : *mce->arguments) {
+                            BUG_CHECK(arg->expression->is<IR::StructExpression>(),
```
Comment (usha1830) [inline backends/dpdk/dpdkArch.h:1154]: Same here

File: backends/dpdk/dpdkArch.h [hunk 3]
```diff
@@ -1135,6 +1135,176 @@ class ElimHeaderCopy : public Transform {
...
+                        }
+                    }
+                }
+            }
+        }
+        return false;
+    }
+};
+
+/// @brief This pass add a pseudo header declaration, it will be used as
```
Comment (usha1830) [inline backends/dpdk/dpdkArch.h:1176]: adds*

File: backends/dpdk/dpdkArch.cpp [hunk 4]
```diff
@@ -2877,4 +2879,174 @@ const IR::Node* ElimHeaderCopy::postorder(IR::Member* m) {
...
+    return st;
+}
+
+bool DpdkAddPseudoHeaderDecl::isHeadersStruct(const IR::Type_Struct* st) {
+    if (!st) return false;
+    auto annon = st->getAnnotation("__packet_data__");
+    if (annon == nullptr) {
+        if (structure->header_type == st->name.name)
+            return true;
+        else {
```
Comment (usha1830) [inline backends/dpdk/dpdkArch.cpp:2918]: The annotations are added early on to mark the main header struct and should be enough to detect this structure.

**Review by usha1830 — CHANGES_REQUESTED** [review]

**Comment by mbudiu-vmw** [comment]
I will assume that @usha1830 will review this. If you need my review please let me know.

File: backends/dpdk/dpdkArch.cpp (hunk 4 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.cpp]: done

File: backends/dpdk/dpdkArch.h (hunk 3 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.h]: done

File: backends/dpdk/dpdkArch.h (hunk 2 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.h]: done

File: backends/dpdk/dpdkArch.cpp (hunk 1 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.cpp]: done

**Review by usha1830 — APPROVED** [review]
Looks fine

**Pull request merged**